### PR TITLE
3587-1

### PR DESCRIPTION
**Pull request by kos1kov**
## Номер
3587

## Номер задания
1

## Ссылка на видео с демо
https://youtu.be/AI2GG-9SU-c

### CODE REVIEW

**Comment by DmitryAvramec** [comment]
Bad commit history, I can't check how you made it because you have all code it 1 commit

File: 3587/fill_output.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    puts table
+  end
+
+  def top_option(top)
+    0.upto(@array_with_info.size - top - 1) { @array_with_info.shift }
+  end
+
+  def best_gem
+    @array_with_info.sort_by! do |element|
+      element[1] * 0.5 + element[2] * 1 + element[3] * 0.7 + element[6] * 0.5 +
```
Comment (DmitryAvramec) [inline 3587/fill_output.rb]: Remove this magic numbers

File: 3587/fill_output.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        element[4] * 0.5 + element[5] * 0.5
+    end
+  end
+
+  def add_text_to_number
```
Comment (DmitryAvramec) [inline 3587/fill_output.rb]: Same here

File: 3587/gem.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+gems:
+ - sinatra
+ - rspec
+ - grape
+ - bullet
+ - capistrano
```
Comment (DmitryAvramec) [inline 3587/gem.yaml]: empty row at the end of file

File: 3587/links_modifier.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  def link_in_homepage
+    link = @ruby_gems_page.links.find do |page_link|
+      page_link.text.include?('Homepage')
+    end
+    link
+  end
+
+  def find_gem_page(gem_name)
+    ruby_gems_link = "https://rubygems.org/gems/#{gem_name}/"
+    page = @agent.get(ruby_gems_link)
```
Comment (DmitryAvramec) [inline 3587/links_modifier.rb]: Move it from this class

File: 3587/links_modifier.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  def initialize(gem_name)
+    @agent = Mechanize.new
+    @ruby_gems_page = find_gem_page(gem_name)
+    @git_page = find_git_page
+    @used_by_page = modifier_url_for_used_by
+  end
+
+  def find_git_page
+    links = link_in_source_code || link_in_homepage
+    git_page = links.click
```
Comment (DmitryAvramec) [inline 3587/links_modifier.rb]: Move it from this class

File: 3587/links_modifier.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def find_gem_page(gem_name)
+    ruby_gems_link = "https://rubygems.org/gems/#{gem_name}/"
+    page = @agent.get(ruby_gems_link)
+    page
+  end
+
+  def modifier_url_for_used_by
+    page = @agent.get("#{@git_page.uri}/network/dependents")
```
Comment (DmitryAvramec) [inline 3587/links_modifier.rb]: Same

File: 3587/scrape_info.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
+require './links_modifier.rb'
+# class that get info about gem from git
+class Scrape
+  attr_reader :watch
+  def initialize(page, used_by_page, name)
+    @page = page
+    @watch = [name]
+    find_info_without_used_by
```
Comment (DmitryAvramec) [inline 3587/scrape_info.rb]: Why you made it in initialize?

File: 3587/scrape_info.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    @used_by_page = used_by_page
+    find_used_by
```
Comment (DmitryAvramec) [inline 3587/scrape_info.rb]: Same here

File: 3587/scrape_info.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def find_forks_stars_watched
```
Comment (DmitryAvramec) [inline 3587/scrape_info.rb]: In ruby we don't use prefixes for methods name like `get_ set_` you just name method with understanding which resource you will get if you call it. Rename it pls

File: 3587/scrape_info.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    temp_array = []
+    @page.css('a.social-count').each do |score|
+      temp_array << score.text.gsub!(/[^\d]/, '').to_i
+    end
+    temp_array
+  end
+
+  def find_contributors
```
Comment (DmitryAvramec) [inline 3587/scrape_info.rb]: Same here

File: 3587/scrape_info.rb [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    contributors = @page.css("span[class ='num text-emphasized']")[3].text.strip
+    contributors = 0 if contributors.empty?
+    contributors
+  end
+
+  def find_issues
```
Comment (DmitryAvramec) [inline 3587/scrape_info.rb]: All of them

File: 3587/top_gems.rb [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+require 'optparse'
+require './links_modifier.rb'
+require './scrape_info.rb'
+require './fill_output.rb'
+# class that contain main info about gem
+class GemStart
+  def initialize
+    @options = {}
+    @array_of_gems = []
+    @parser = OptionParser.new do |opt|
```
Comment (DmitryAvramec) [inline 3587/top_gems.rb]: Why I can see this in initializer? You can send Option parser object as an argument for GemStart initializer

File: 3587/top_gems.rb [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+      opt.on('--name NAME')
+      opt.on('--file FILE')
+    end
+    @parser.parse!(into: @options)
+    @main_array = []
+    run
+  end
+
+  def user_option
+    file = @options[:file] || 'gem.yaml'
```
Comment (DmitryAvramec) [inline 3587/top_gems.rb]: This is Not this class logic

File: 3587/top_gems.rb [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+  def run
+    file, name, top = user_option
+    get_gem_from_file(file, name)
+    parse
+    FillOutput.new(@main_array, top)
+  end
+
+  def get_gem_from_file(file_name, gem_name)
+    file = File.open(file_name)
```
Comment (DmitryAvramec) [inline 3587/top_gems.rb]: ????? extension for file must be YAML

**Comment by DmitryAvramec** [comment]
- [ ] Other flags? Combinations of flags?

- [ ] Single responcibility for clasess

- [ ] Fix commit history and commit names

- [ ] Remove all disables

**Comment by kos1kov** [comment]
- [x] Other flags? Combinations of flags?

- [x] Single responcibility for clasess

- [x] Fix commit history and commit names

- [x] Remove all disables

File: 3587/analize_stats.rb [hunk 15]
```diff
@@ -0,0 +1,18 @@
+# class that analize gems_stats
+class AnalizeGems
```
Comment (DmitryAvramec) [inline 3587/analize_stats.rb]: Hm, analize is a verb?

File: 3587/scrape_info.rb [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+  def contributors
+    @contributors ||= page.css(CONTRIBUTORS_CSS_CLASS)[3].text.strip
+  end
+
+  def issues
+    @issues ||= page.at_css('.Counter').text
+  end
+
+  def used_by_count
```
Comment (Moleved) [inline 3587/scrape_info.rb]: amount `used_by_amount`

File: 3587/analize_stats.rb [hunk 17]
```diff
@@ -0,0 +1,21 @@
+# class that analyze gems_stats
+class AnalyzerGems
```
Comment (Moleved) [inline 3587/analize_stats.rb]: Don't use verbs for class naming it should be noun

File: 3587/fill_output.rb [hunk 18]
```diff
@@ -0,0 +1,28 @@
+require 'terminal-table'
+require './fill_row.rb'
+# class about output info
+class FillOutput
```
Comment (Moleved) [inline 3587/fill_output.rb]: Should be noun

**Comment by Moleved** [comment]
Add text to README or remove it

File: 3587/gem_page.rb [hunk 19]
```diff
@@ -0,0 +1,42 @@
+# class for getting links
```
Comment (DmitryAvramec) [inline 3587/gem_page.rb]: https://rubizza-camp.slack.com/archives/CKN3DH9C2/p1563224295438400

File: 3587/lib/options.rb [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+  def call
+    parser
+    @arg
+  end
+
+  private
+
+  def parser
+    @parser = OptionParser.new do |opt|
+      opt.on('--top TOP')
```
Comment (DmitryAvramec) [inline 3587/lib/options.rb]: some description mb?

File: 3587/lib/scrape_info.rb [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+
+  def stars_amount
+    main_amounts[1]
+  end
+
+  def forks_amount
+    main_amounts[0]
+  end
+
+  def main_amounts
```
Comment (DmitryAvramec) [inline 3587/lib/scrape_info.rb]: Mb make hash here? why you have 3 methods with magic numbers?

Comment (DmitryAvramec) [reply 3587/lib/scrape_info.rb]: And rename it

File: 3587/lib/scrape_info.rb [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+
+  def contributors
+    page.css(CONTRIBUTORS_CSS_CLASS)[3].text.strip.to_i
+  end
+
+  def issues
+    page.at_css('.Counter').text.to_i
+  end
+
+  def used_by_amount
```
Comment (DmitryAvramec) [inline 3587/lib/scrape_info.rb]: again `amount` =\\

File: 3587/lib/top_gems.rb [hunk 23]
```diff
@@ -0,0 +1,20 @@
+class GemStart
+  def initialize
+    @main_hash = {}
+  end
+
+  def parse(array_of_gems)
+    array_of_gems.each do |gem|
```
Comment (DmitryAvramec) [inline 3587/lib/top_gems.rb]: each_with_object

File: 3587/lib/top_gems.rb [hunk 24]
```diff
@@ -0,0 +1,20 @@
+class GemStart
+  def initialize
+    @main_hash = {}
```
Comment (DmitryAvramec) [inline 3587/lib/top_gems.rb]: useless

File: 3587/lib/fill_output.rb [hunk 25]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  def sort
+    @gem_hash = @gem_hash.sort_by { |_key, value| -value[:score] }.to_h
+  end
+
+  def print
+    sort
+    table = Terminal::Table.new
+    how_many_print(@gem_hash.size).times do |index|
```
[the remark this answers is not in the document]
Comment (kos1kov) [reply 3587/lib/fill_output.rb]: not found other solution for integer

File: 3587/lib/options.rb (hunk 20 above)
Comment (DmitryAvramec) [reply 3587/lib/options.rb]: https://ruby-doc.org/stdlib-2.6.1/libdoc/optparse/rdoc/OptionParser.html

File: 3587/lib/scrape_info.rb [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+      fork: forks,
+      used_by: used_by.gsub!(/[^\d]/, '').to_i,
+      contributors: contributors, # Move to_i method to contributors
+      issues: issues
+    }
+  end
+
+  private
+
+  def watch
```
Comment (DmitryAvramec) [inline 3587/lib/scrape_info.rb]: It's 1 watch? Blackwatch))))

File: 3587/lib/top_gems.rb [hunk 27]
```diff
@@ -0,0 +1,20 @@
+class GemStart
```
Comment (DmitryAvramec) [inline 3587/lib/top_gems.rb]: bad class name

File: 3587/lib/top_gems.rb [hunk 28]
```diff
@@ -0,0 +1,20 @@
...
+  attr_reader :options
+
+  def initialize
+    @options = ParseOption.call
+  end
+
+  def parse
```
Comment (DmitryAvramec) [inline 3587/lib/top_gems.rb]: better to name it `perform` and call it from some kind of 
```ruby
def self.perform
  # Any kind of preparation here
  new.perform # If you don't need to send arguments to initializer
end

def perform
  # Some actions here
  # Parse
  # output
end
```

File: 3587/lib/fill_row.rb [hunk 29]
```diff
@@ -0,0 +1,20 @@
+class TableRow
+  attr_reader :name, :stats
+
+  def initialize(name, stats)
+    @name = name
+    @stats = stats
+  end
+
+  def table_row
```
Comment (Moleved) [inline 3587/lib/fill_row.rb]: `TableRow.new.table_row` to call table row from table row?

File: 3587/lib/gem_list.rb [hunk 30]
```diff
@@ -0,0 +1,16 @@
+class GemList
+  def initialize(options)
+    @file = options[:file]
+    @argument_name = options[:name]
+  end
+
+  def hash_of_gem
```
Comment (Moleved) [inline 3587/lib/gem_list.rb]: Should be private

File: 3587/lib/gem_page.rb [hunk 31]
```diff
@@ -0,0 +1,38 @@
...
+
+  def initialize(gem_name)
+    @agent = Mechanize.new
+    @ruby_gems_page = gem_page(gem_name)
+    @github_page = git_page
+    @used_by_page = page_used_by
+  end
+
+  def git_page
+    links = link_in_source_code || link_in_homepage
```
[the remark this answers is not in the document]
Comment (Moleved) [reply 3587/lib/gem_page.rb]: Fix comment

File: 3587/lib/options.rb [hunk 32]
```diff
@@ -0,0 +1,31 @@
...
+
+  def call
+    parser
+    @arg
+  end
+
+  private
+
+  def parser
+    @parser = OptionParser.new do |opt|
```
Comment (Moleved) [inline 3587/lib/options.rb]: Useless instance variable

File: 3587/lib/options.rb [hunk 33]
```diff
@@ -0,0 +1,31 @@
...
+      opt.on('--top TOP', 'Number of items displayed')
+      opt.on('--name NAME', 'Word that find in gems name')
+      opt.on('--file FILE', 'Path to file')
+    end.parse!(into: arg)
+    arg
```
Comment (Moleved) [inline 3587/lib/options.rb]: You no need returning arg from parser

File: 3587/lib/options.rb [hunk 34]
```diff
@@ -0,0 +1,31 @@
...
+    }
+  end
+
+  def self.call
+    new.call
+  end
+
+  def call
+    parser
+    @arg
```
[the remark this answers is not in the document]
Comment (Moleved) [reply 3587/lib/options.rb]: `arg`

File: 3587/lib/top_gems.rb [hunk 35]
```diff
@@ -0,0 +1,29 @@
...
+
+  def perform
+    array_of_gems = GemList.new(options).list_of_gem
+    hash_gem = parse(array_of_gems)
+    Output.new(hash_gem, options[:top]).print
+  end
+
+  def parse(array_of_gems)
+    array_of_gems.each_with_object({}) do |gem, hash|
+      begin
```
[the remark this answers is not in the document]
Comment (kos1kov) [reply 3587/lib/top_gems.rb]: Sorry, this error not displaed at my local rubucop

File: 3587/lib/top_gems.rb [hunk 36]
```diff
@@ -0,0 +1,27 @@
...
+    hash_gem = parse(array_of_gems)
+    Output.new(hash_gem, options[:top]).print
+  end
+
+  def parse(array_of_gems)
+    array_of_gems.each_with_object({}) do |gem, hash|
+        hash_with_stat = Scraper.new(gem).hash_with_info_about_gem
+        hash_with_stat[:score] = AnalyzerGems.new(hash_with_stat).score_of_gems
+        hash[gem] = hash_with_stat
+      rescue NoMethodError
```
[the remark this answers is not in the document]
Comment (kos1kov) [reply 3587/lib/top_gems.rb]: some here

File: 3587/1/lib/gem_page.rb [hunk 37]
```diff
@@ -0,0 +1,34 @@
...
+
+  def link_in_source_code
+    @ruby_gems_page.links.find do |page_link|
+      page_link.text.include?('Source Code')
+    end
+  end
+
+  def gem_page(gem_name)
+    @agent.get(URI.join(RUBYGEMS_URL, gem_name))
+  rescue Mechanize::ResponseCodeError
```
[the remark this answers is not in the document]
Comment (kos1kov) [reply 3587/1/lib/gem_page.rb]: Why, when we try to get https://rubygems.org/gems/blablabla, we get error that page not found

**Pull request merged**